compat/src/suspense-list.js

```diff
@@ -1,150 +1,127 @@
 	if (
-		parent.type.name === SuspenseList.name &&
-		parent._component.__getRevealOrder() === 't'
+		!list.props.revealOrder ||
+		(list.props.revealOrder[0] === 't' && list._map.size)
 	) {
-		order = 't';
-	}
-
-	if (!order) {
-		return '';
+		return;
 	}
 
-	return order[0];
-};
-
-SuspenseList.prototype.__suspenseDidResolve = function(vnode) {
-	this._suspenseBoundaries.some((suspenseBoundary, index) => {
-		if (suspenseBoundary === vnode) {
-			const nextVNode = this._suspenseBoundaries[index + 1];
-			const cb = nextVNode ? this._suspenseCallbacks.get(nextVNode) : null;
-			if (nextVNode && cb && !nextVNode._component._isSuspenseResolved) {
-				cb();
-			}
-			return true;
+	// Walk the currently suspended children in order, calling their
+	// stored callbacks on the way. Stop if we encounter a child that
+	// has not been completely resolved yet.
+	node = list._next;
+	while (node) {
+		while (node.length > 3) {
+			node.pop()();
 		}
-	});
-	if (
-		this._suspenseBoundaries.every(
-			boundary => boundary._component._isSuspenseResolved
-		)
-	) {
-		this._isSuspenseResolved = true;
-		suspenseDidResolve(this._vnode);
+		if (node[RESOLVED_COUNT] < node[SUSPENDED_COUNT]) {
+			break;
+		}
+		list._next = node = node[NEXT_NODE];
 	}
 };
 
-SuspenseList.prototype.__suspenseWillResolve = function(vnode, cb) {
-	this._suspenseCallbacks.set(vnode, cb);
+// Things we do here to save some bytes but are not proper JS inheritance:
+// - call `new Component()` as the prototype
+// - do not set `Suspense.prototype.constructor` to `Suspense`
+SuspenseList.prototype = new Component();
+
+SuspenseList.prototype._suspended = function(child) {
+	const list = this;
+	const delegated = suspended(list._vnode);
+
+	let node = list._map.get(child);
+	node[SUSPENDED_COUNT]++;
 
-	/**
-	 * A Suspense list with revealorder=t is ready render only when all
-	 * of its children are ready to render.
-	 */
-	if (this.__getRevealOrder() === 't') {
-		if (
-			this._suspenseBoundaries.every(
-				suspenseBoundary =>
-					this._suspenseCallbacks.has(suspenseBoundary) ||
-					suspenseBoundary._component._isSuspenseResolved
-			)
-		) {
-			return suspenseWillResolve(this._vnode, () => {
-				this._readyToRender = true;
-				this.__findAndResolveNextcandidate();
-			});
+	return unsuspend => {
+		const wrappedUnsuspend = () => {
+			if (!list.props.revealOrder) {
+				// Special case the undefined (falsy) revealOrder, as there
+				// is no need to coordinate a specific order or unsuspends.
+				unsuspend();
+			} else {
+				node.push(unsuspend);
+				resolve(list, child, node);
+			}
+		};
+		if (delegated) {
+			delegated(wrappedUnsuspend);
+		} else {
+			wrappedUnsuspend();
 		}
-	} else {
-		this.__findAndResolveNextcandidate();
-	}
+	};
 };
 
-SuspenseList.prototype.__findAndResolveNextcandidate = function() {
-	if (!this._readyToRender) {
-		return;
-	}
+SuspenseList.prototype.render = function(props) {
+	this._next = null;
+	this._map = new Map();
```

Comment:
`this._map =this._map || new Map();`

Dont re-initialize the map here.
I haven't been able to find a way to write test for this but here the use case.

All the Suspensions are completed with revealOrder=forwards. And then one more enters on 2nd positions.
This might put completed suspensions(from third positions onwards) back in loading state(atleast earlier version of suspense did).

Reply:
Also make sure that in above use case we should not re-call the `unsuspend` method of children as sometimes the render is not a pure function in real world.

in my PR `_isSuspensionCompleted` takes care of this